Improve triton bsr_dense_mm performance on column-major ordered inputs with float32 dtype

pearu opened the pull request:
As in the title.

The bsr_dense_mm performance on inputs using column-major storage order is relevant for `linear(x, W)` operation that for BSR weights is defined as `bsr_dense_mm(W, x.transpose(-2, -1)).transpose(-2, 1)` so that the second argument to `bse_dense_mm` is a strided tensor using column-major storage order when `x` is C-contiguous.

For large inputs (size > 1000) and moderate sparsity in the BSR input, the speed up can be more than 3 times, as illustrated in the following figure (raw data: [bench_bsr_dense_mm_1_results.txt](https://github.com/pytorch/pytorch/files/12512245/bench_bsr_dense_mm_1_results.txt)):

![bench_bsr_dense_mm_1](https://github.com/pytorch/pytorch/assets/402156/c6372008-dfae-4d26-b119-2c3c944a74ae)

For small inputs (size=512), there exists a slight degradation of performance.

For row-major ordered inputs, there is no change in performance (see raw data above).

For inputs with float16 dtype, there is no considerable change in performance (see blue marks in the figure).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108601
* __->__ #108512

pearu committed:
Improve triton bsr_dense_mm performance on column-major ordered inputs

[ghstack-poisoned]

pearu committed:
Update on "Improve triton bsr_dense_mm performance on column-major ordered inputs"




[ghstack-poisoned]

pearu commented on torch/sparse/_triton_ops.py:
```diff
@@ -82,7 +82,7 @@ def is_compatible_blocksize(b):
 
 def make_triton_contiguous(t):
-    if t.stride(-2) > 1 and t.stride(-1) > 1:
+    if (t.stride(-2) > 1 or t.dtype is torch.float32) and t.stride(-1) > 1:
```
This is the gist of the PR: apparently, using contiguous input to `triton.language.dot` pays off the overheads from `contiguous()` call when inputs have float32 dtype and relatively large sizes (>1000).

cpuhrsch commented on torch/sparse/_triton_ops.py:
```diff
@@ -459,7 +462,7 @@ def _bsr_strided_dense_rowspace_kernel(
 
             # do block mm
-            output_acc_block += tl.dot(values_block, dense_block, allow_tf32=allow_tf32)
+            output_acc_block += tl.dot(values_block, dense_block, allow_tf32=allow_tf32, out_dtype=acc_dtype)
```
This feels like a significant bugfix

cpuhrsch replied:
Line 337 above has `acc_block += tl.dot(mat1_block, mat2_block, allow_tf32=allow_tf32)` and also doesn't pass the out_dtype.

cpuhrsch replied:
Could the change in performance come from adjusting the out_dtype here? Maybe if so the call to contiguous isn't what's needed, but a different accumulation dtype which in turn influences which instruction is used to perform the dot.

pearu replied:
> This feels like a significant bugfix

Actually, it is not too bad: the default dtype for `tl.dot` output is float32 which coincides with the accumulator dtype for bfloat16/float16 inputs. So, the fix is effective only for float32 input type that uses float64 accumulator dtype.

> Line 337 above... does not specify `out_dtype`.

Fixed.

> Could the change in performance come from adjusting the out_dtype here?

No: Disabling `out_dtype` produces the same performance improvement figure as in the description.

> Maybe if so the call to contiguous isn't what's needed...

That would be nice indeed but unfortunately, it is not the case here. Notice that the performance improvement for float32 inputs is present for larger input and block sizes and I would attribute this effect to minimizing cache misses (or similar phenomena) when operating on contiguous inputs (although, the cache misses are not directly measurable here). For smaller input sizes, cache misses are less likely and the overhead of the `contiguous()` call becomes apparent in a slight reduction of the performance.

pearu replied:
In fact, considering
https://github.com/openai/triton/blob/36859aebff1897f3bdf6dd19ba775e5961848943/python/triton/language/semantic.py#L1294C1-L1305C34
`out_dtype` value is in many cases just ignored by triton and specifying `out_dtype` for `float32` inputs has no effect.

pearu committed:
Update on "Improve triton bsr_dense_mm performance on column-major ordered inputs with float32 dtype"

As in the title.

The bsr_dense_mm performance on inputs using column-major storage order is relevant for `linear(x, W)` operation that for BSR weights is defined as `bsr_dense_mm(W, x.transpose(-2, -1)).transpose(-2, 1)` so that the second argument to `bse_dense_mm` is a strided tensor using column-major storage order when `x` is C-contiguous.

For large inputs (size > 1000) and moderate sparsity in the BSR input, the speed up can be more than 3 times, as illustrated in the following figure (raw data: [bench_bsr_dense_mm_1_results.txt](https://github.com/pytorch/pytorch/files/12512245/bench_bsr_dense_mm_1_results.txt)):

![bench_bsr_dense_mm_1](https://github.com/pytorch/pytorch/assets/402156/c6372008-dfae-4d26-b119-2c3c944a74ae)

For small inputs (size=512), there exists a slight degradation of performance.

For row-major ordered inputs, there is no change in performance (see raw data above).

For inputs with float16 dtype, there is no considerable change in performance (see blue marks in the figure).




[ghstack-poisoned]

pearu committed:
Update on "Improve triton bsr_dense_mm performance on column-major ordered inputs with float32 dtype"

As in the title.

The bsr_dense_mm performance on inputs using column-major storage order is relevant for `linear(x, W)` operation that for BSR weights is defined as `bsr_dense_mm(W, x.transpose(-2, -1)).transpose(-2, 1)` so that the second argument to `bse_dense_mm` is a strided tensor using column-major storage order when `x` is C-contiguous.

For large inputs (size > 1000) and moderate sparsity in the BSR input, the speed up can be more than 3 times, as illustrated in the following figure (raw data: [bench_bsr_dense_mm_1_results.txt](https://github.com/pytorch/pytorch/files/12512245/bench_bsr_dense_mm_1_results.txt)):

![bench_bsr_dense_mm_1](https://github.com/pytorch/pytorch/assets/402156/c6372008-dfae-4d26-b119-2c3c944a74ae)

For small inputs (size=512), there exists a slight degradation of performance.

For row-major ordered inputs, there is no change in performance (see raw data above).

For inputs with float16 dtype, there is no considerable change in performance (see blue marks in the figure).




[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge
